provider/cluster/hostname.go

```diff
@@ -4,79 +4,174 @@ import (
 	"context"
 	"fmt"
 	lifecycle "github.com/boz/go-lifecycle"
+	cosmostypes "github.com/cosmos/cosmos-sdk/types"
 	dtypes "github.com/ovrclk/akash/x/deployment/types"
 	"github.com/pkg/errors"
 	"strings"
 	"sync"
 )
 
-type reserveRequest struct {
-	hostnames []string
-	result    chan<- error
-	doReserve bool
-	dID       dtypes.DeploymentID
-}
-
 type HostnameServiceClient interface {
-	ReserveHostnames(hostnames []string, did dtypes.DeploymentID) <-chan error
-	ReleaseHostnames(hostnames []string)
-	CanReserveHostnames(hostnames []string, did dtypes.DeploymentID) <-chan error
+	ReserveHostnames(hostnames []string, dID dtypes.DeploymentID) ReservationResult
+	ReleaseHostnames(dID dtypes.DeploymentID)
+	CanReserveHostnames(hostnames []string, ownerAddr cosmostypes.Address) <-chan error
 }
 
 type SimpleHostnames struct {
 	Hostnames map[string]dtypes.DeploymentID
 	lock      sync.Mutex
 } /* Used in test code */
 
-func (sh *SimpleHostnames) ReserveHostnames(hostnames []string, did dtypes.DeploymentID) <-chan error {
+type ReplacedHostname struct {
+	Hostname                   string
+	PreviousDeploymentSequence uint64
+}
+
+type ReservationResult struct {
+	ChErr               <-chan error
+	ChReplacedHostnames <-chan []ReplacedHostname
+}
+
+func (rr ReservationResult) Wait(wait <-chan struct{}) ([]ReplacedHostname, error) {
+	select {
+	case err := <-rr.ChErr:
+		if err != nil {
+			return nil, err
+		}
+	case <-wait:
+		return nil, errors.New("bob")
+	}
+
+	return <-rr.ChReplacedHostnames, nil
```

Comment:
this can/should be in the `select` so that you don't have to worry about closing or sending `nil` to `ChErr`.